File: vllm/attention/ops/cpac_ops.py
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
"""
Python bindings for CPAC CUDA operations
"""

import torch
from typing import Tuple, Optional

# Try to import the compiled CUDA extensions
try:
    import vllm._C
    # Check if CPAC ops are available in the compiled extension
    if hasattr(vllm._C, 'cpac_ops'):
        cpac_ops = vllm._C.cpac_ops
        HAS_CPAC_OPS = True
    else:
        HAS_CPAC_OPS = False
except ImportError:
    HAS_CPAC_OPS = False
    

def compute_page_features(
    page_data: torch.Tensor,
    feature_dim: int = 128
) -> torch.Tensor:
    """
    Compute features for a page to be used in similarity matching
    
    Args:
        page_data: Tensor of shape [num_heads, head_size, block_size]
        feature_dim: Dimension of output feature vector
    
    Returns:
        features: Tensor of shape [feature_dim]
    """
    if HAS_CPAC_OPS and page_data.is_cuda:
        return cpac_ops.compute_page_features_cuda(page_data, feature_dim)
    else:
        # CPU fallback implementation
        num_heads, head_size, block_size = page_data.shape
        features = []
        
        # Mean and variance per head
        reshaped = page_data.view(num_heads, -1)
        mean_per_head = reshaped.mean(dim=1)
        var_per_head = reshaped.var(dim=1)
        features.append(mean_per_head)
        features.append(var_per_head.sqrt())
        
        # Attention pattern signature
        pattern_sig = page_data.abs().mean(dim=1).flatten()
        features.append(pattern_sig[:min(pattern_sig.size(0), feature_dim - 2*num_heads)])
        
        # Pad if necessary
        feature_tensor = torch.cat(features)
        if feature_tensor.size(0) < feature_dim:
            padding = torch.zeros(feature_dim - feature_tensor.size(0), 
                                device=page_data.device, dtype=torch.float32)
            feature_tensor = torch.cat([feature_tensor, padding])
        
        return feature_tensor[:feature_dim]


def compute_similarity(
    features1: torch.Tensor,
    features2: torch.Tensor
) -> torch.Tensor:
    """
    Compute cosine similarity between two feature vectors
    
    Args:
        features1: Feature vector 1
        features2: Feature vector 2
    
    Returns:
        similarity: Scalar similarity value
    """
    if HAS_CPAC_OPS and features1.is_cuda:
        return cpac_ops.compute_similarity_cuda(features1, features2)
    else:
        # CPU fallback
        return torch.nn.functional.cosine_similarity(
            features1.unsqueeze(0), features2.unsqueeze(0)
        )


def compress_delta(
    target_page: torch.Tensor,
    base_page: torch.Tensor,
    delta_bits: int = 8
) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Compress target page as delta from base page
    
    Args:
        target_page: Page to compress
        base_page: Base page for delta encoding
        delta_bits: Number of bits for quantized delta
    
    Returns:
        quantized_delta: Quantized delta tensor
        scale: Scale factor for dequantization
    """
    if HAS_CPAC_OPS and target_page.is_cuda:
        return cpac_ops.compress_delta_cuda(target_page, base_page, delta_bits)
    else:
        # CPU fallback
        delta = target_page - base_page
        max_val = delta.abs().max().item()
        
        if max_val == 0:
            return torch.zeros_like(delta, dtype=torch.int8), torch.tensor(1.0)
        
        max_int = (1 << (delta_bits - 1)) - 1
        scale = max_val / max_int
        
        quantized = torch.round(delta / scale).clamp(-max_int, max_int).to(torch.int8)
        
        return quantized, torch.tensor(scale)


def decompress_delta(
    quantized_delta: torch.Tensor,
    base_page: torch.Tensor,
    scale: float
) -> torch.Tensor:
    """
    Decompress page from quantized delta and base page
    
    Args:
        quantized_delta: Quantized delta tensor
        base_page: Base page for reconstruction
        scale: Scale factor from compression
    
    Returns:
        decompressed_page: Reconstructed page
    """
    if HAS_CPAC_OPS and quantized_delta.is_cuda:
        return cpac_ops.decompress_delta_cuda(quantized_delta, base_page, scale)
    else:
        # CPU fallback
        delta = quantized_delta.float() * scale
        return base_page + delta.to(base_page.dtype)


def batch_compute_similarities(
    features_list: torch.Tensor,
    top_k: int = 4
) -> torch.Tensor:
    """
    Compute pairwise similarities for a batch of feature vectors
    
    Args:
        features_list: Tensor of shape [num_pages, feature_dim]
        top_k: Number of top similar pages to return per page
    
    Returns:
        top_similarities: Tensor of shape [num_pages, top_k] with similarity scores
        top_indices: Tensor of shape [num_pages, top_k] with page indices
    """
    num_pages = features_list.size(0)
    
    # Compute pairwise similarities
    similarities = torch.mm(features_list, features_list.t())
    
    # Normalize by norms
    norms = torch.norm(features_list, dim=1, keepdim=True)
    similarities = similarities / (norms * norms.t() + 1e-8)
    
    # Mask out self-similarities
    mask = torch.eye(num_pages, device=features_list.device)
    similarities = similarities * (1 - mask) - mask
    
    # Get top-k similarities
    top_similarities, top_indices = torch.topk(similarities, k=min(top_k, num_pages-1), dim=1)
    
    return top_similarities, top_indices


def estimate_compression_ratio(
    target_page: torch.Tensor,
    base_page: torch.Tensor,
    delta_bits: int = 8
) -> float:
    """
    Estimate compression ratio without actually compressing
    
    Args:
        target_page: Page to compress
        base_page: Base page for delta encoding
        delta_bits: Number of bits for quantized delta
    
    Returns:
        estimated_ratio: Estimated compression ratio
    """
    original_bits = target_page.numel() * target_page.element_size() * 8
    compressed_bits = target_page.numel() * delta_bits + 32  # +32 for scale
    
    # Estimate quality loss
    delta = target_page - base_page
    max_val = delta.abs().max().item()
    
    if max_val == 0:
        return float('inf')  # Perfect compression
    
    max_int = (1 << (delta_bits - 1)) - 1
    scale = max_val / max_int
    quantization_error = scale / 2  # Average quantization error
    
    # Penalize high quantization error
    quality_factor = 1.0 / (1.0 + quantization_error)
    
    return (original_bits / compressed_bits) * quality_factor


class CPACKernel:
    """
    Wrapper class for CPAC CUDA kernels with automatic fallback
    """
    
    def __init__(self):
        self.has_cuda = HAS_CPAC_OPS
        if not self.has_cuda:
            import warnings
            warnings.warn("CPAC CUDA kernels not available, using CPU fallback")
    
    def compute_features(self, page_data: torch.Tensor, feature_dim: int = 128) -> torch.Tensor:
        return compute_page_features(page_data, feature_dim)
    
    def compute_similarity(self, feat1: torch.Tensor, feat2: torch.Tensor) -> torch.Tensor:
        return compute_similarity(feat1, feat2)
    
    def compress(self, target: torch.Tensor, base: torch.Tensor, 
                 bits: int = 8) -> Tuple[torch.Tensor, torch.Tensor]:
        return compress_delta(target, base, bits)
    
    def decompress(self, delta: torch.Tensor, base: torch.Tensor, 
                   scale: float) -> torch.Tensor:
        return decompress_delta(delta, base, scale)


# Global kernel instance
cpac_kernel = CPACKernel()

File: csrc/cpac_ops.cpp
#include <torch/extension.h>
#include <vector>

// Forward declarations of CUDA functions
torch::Tensor compute_page_features_cuda(torch::Tensor page_data, int feature_dim);
torch::Tensor compute_similarity_cuda(torch::Tensor features1, torch::Tensor features2);
std::tuple<torch::Tensor, torch::Tensor> compress_delta_cuda(
    torch::Tensor target_page, torch::Tensor base_page, int delta_bits);
torch::Tensor decompress_delta_cuda(
    torch::Tensor quantized_delta, torch::Tensor base_page, float scale);

// Wrapper functions with error checking
torch::Tensor cpac_compute_page_features(torch::Tensor page_data, int feature_dim) {
    TORCH_CHECK(page_data.is_cuda(), "page_data must be a CUDA tensor");
    TORCH_CHECK(page_data.dim() == 3, "page_data must be 3-dimensional");
    TORCH_CHECK(feature_dim > 0, "feature_dim must be positive");
    
    return compute_page_features_cuda(page_data, feature_dim);
}

torch::Tensor cpac_compute_similarity(torch::Tensor features1, torch::Tensor features2) {
    TORCH_CHECK(features1.is_cuda(), "features1 must be a CUDA tensor");
    TORCH_CHECK(features2.is_cuda(), "features2 must be a CUDA tensor");
    TORCH_CHECK(features1.size(0) == features2.size(0), 
                "features1 and features2 must have the same size");
    
    return compute_similarity_cuda(features1, features2);
}

std::tuple<torch::Tensor, torch::Tensor> cpac_compress_delta(
    torch::Tensor target_page, torch::Tensor base_page, int delta_bits) {
    TORCH_CHECK(target_page.is_cuda(), "target_page must be a CUDA tensor");
    TORCH_CHECK(base_page.is_cuda(), "base_page must be a CUDA tensor");
    TORCH_CHECK(target_page.sizes() == base_page.sizes(), 
                "target_page and base_page must have the same shape");
    TORCH_CHECK(delta_bits > 0 && delta_bits <= 16, 
                "delta_bits must be between 1 and 16");
    
    return compress_delta_cuda(target_page, base_page, delta_bits);
}

torch::Tensor cpac_decompress_delta(
    torch::Tensor quantized_delta, torch::Tensor base_page, float scale) {
    TORCH_CHECK(quantized_delta.is_cuda(), "quantized_delta must be a CUDA tensor");
    TORCH_CHECK(base_page.is_cuda(), "base_page must be a CUDA tensor");
    TORCH_CHECK(quantized_delta.dtype() == torch::kInt8, 
                "quantized_delta must be int8");
    TORCH_CHECK(scale > 0, "scale must be positive");
    
    return decompress_delta_cuda(quantized_delta, base_page, scale);
}

// Registration function to be called from torch_bindings.cpp
void register_cpac_ops(py::module& m) {
    // Create submodule for CPAC operations
    auto cpac_module = m.def_submodule("cpac_ops", "CPAC compression operations");
    
    cpac_module.def(
        "compute_page_features",
        &cpac_compute_page_features,
        "Compute page features for similarity matching",
        py::arg("page_data"),
        py::arg("feature_dim") = 128
    );
    
    cpac_module.def(
        "compute_similarity",
        &cpac_compute_similarity,
        "Compute cosine similarity between feature vectors",
        py::arg("features1"),
        py::arg("features2")
    );
    
    cpac_module.def(
        "compress_delta",
        &cpac_compress_delta,
        "Compress page using delta encoding",
        py::arg("target_page"),
        py::arg("base_page"),
        py::arg("delta_bits") = 8
    );
    
    cpac_module.def(
        "decompress_delta",
        &cpac_decompress_delta,
        "Decompress page from delta encoding",
        py::arg("quantized_delta"),
        py::arg("base_page"),
        py::arg("scale")
    );
}